File: scripts/_prepare-demo-cad.py
#!/usr/bin/env python3
"""
Helper that runs INSIDE the cad-converter container to batch-convert all STEP
files in /work/step/ into GLB (/work/glb/) + PNG thumbnails (/work/thumbnails/).

Invoked by scripts/prepare-demo-cad.ts. Requires Xvfb running on DISPLAY=:99
for thumbnail rendering — the orchestrator handles that.

Honors:
  FORCE=1        re-convert files that already have a GLB
  ONLY=str       only process basenames containing the substring
  SKIP_THUMBS=1  leave existing thumbnails alone (geometry-only re-bake)
"""

from __future__ import annotations

import json
import os
import sys
import time
import traceback
from pathlib import Path

WORK_DIR = Path(sys.argv[1] if len(sys.argv) > 1 else "/work")
STEP_DIR = WORK_DIR / "step"
GLB_DIR = WORK_DIR / "glb"
THUMB_DIR = WORK_DIR / "thumbnails"
TMP_STL_DIR = Path("/tmp/demo-stl")

FORCE = os.environ.get("FORCE", "0") == "1"
ONLY = os.environ.get("ONLY", "")
SKIP_THUMBS = os.environ.get("SKIP_THUMBS", "0") == "1"

# One JSON per GLB, naming the parts inside it — looked up by the same
# `cadFileBase` key as glb/ and thumbnails/, so the seed needs no index.
# Written because the converter now returns them and the app needs them: an
# assembly GLB carries a glTF node per leaf part, and
# `vault_files.cad_metadata.nodes` is what tells the viewer a model can be
# taken apart. A model with no parts to select gets `{"nodes": []}` rather than
# no file, so a missing file means "not converted yet" and not "single part".
NODE_DIR = WORK_DIR / "nodes"

GLB_DIR.mkdir(parents=True, exist_ok=True)
THUMB_DIR.mkdir(parents=True, exist_ok=True)
NODE_DIR.mkdir(parents=True, exist_ok=True)
TMP_STL_DIR.mkdir(parents=True, exist_ok=True)

# Make the cad_converter package importable.
sys.path.insert(0, "/app/src")

from cad_converter.converter import convert_single_with_colors  # noqa: E402
from cad_converter.models import MeshQuality  # noqa: E402

steps = sorted(STEP_DIR.glob("*.step"))
if ONLY:
    steps = [s for s in steps if ONLY.lower() in s.stem.lower()]

print(f"[helper] {len(steps)} STEP files (FORCE={FORCE}, ONLY={ONLY!r})", flush=True)

ok = 0
skipped = 0
failed: list[tuple[str, str]] = []

t0 = time.time()
for i, step in enumerate(steps, start=1):
    base = step.stem
    glb_out = GLB_DIR / f"{base}.glb"
    thumb_out = THUMB_DIR / f"{base}.png"
    node_out = NODE_DIR / f"{base}.json"
    stl_out = TMP_STL_DIR / f"{base}.stl"

    if not FORCE and glb_out.exists() and thumb_out.exists() and node_out.exists():
        skipped += 1
        continue

    # Re-rendering a thumbnail from unchanged geometry produces a different
    # PNG for the same picture, which is churn in a 197 MB dataset for nothing.
    want_thumb = None if (SKIP_THUMBS and thumb_out.exists()) else str(thumb_out)

    print(f"[{i:>3}/{len(steps)}] {base}", flush=True)
    try:
        result = convert_single_with_colors(
            str(step),
            str(stl_out),
            str(glb_out),
            MeshQuality.STANDARD,
            True,
            want_thumb,
        )
        if glb_out.exists():
            nodes = [n.model_dump() for n in result.glb_nodes]
            node_out.write_text(
                json.dumps({"nodes": nodes}, indent=1), encoding="utf-8"
            )
            print(f"        {len(nodes)} selectable part(s)", flush=True)
            ok += 1
        else:
            failed.append((base, "GLB not produced"))
        # STLs aren't shipped — discard.
        try:
            stl_out.unlink(missing_ok=True)
        except Exception:
            pass
    except Exception as e:
        failed.append((base, f"{type(e).__name__}: {e}"))
        traceback.print_exc()

elapsed = time.time() - t0
print()
print(f"[helper] done in {elapsed:.1f}s — ok={ok}, skipped={skipped}, failed={len(failed)}", flush=True)
if failed:
    print("[helper] failures:", flush=True)
    for base, why in failed:
        print(f"   {base}: {why}", flush=True)

# Exit success even with some failures — the seed handles missing assets gracefully.
sys.exit(0)

File: scripts/prepare-demo-cad.ts
/**
 * Generate pre-converted GLB + PNG thumbnail for every STEP in robot-arm/step/.
 *
 * One-time tooling. Pulls the published cad-converter image if needed, then runs
 * a single container that iterates the STEP directory and writes GLB + thumbnail
 * outputs back into the dataset tree.
 *
 * Run with:
 *   tsx scripts/prepare-demo-cad.ts
 *
 * Env:
 *   CAD_IMAGE    override image ref (default: the published cascadia-cad-converter)
 *   FORCE        set to 1 to re-convert files that already have a GLB
 *   ONLY         substring filter on STEP basename (e.g., ONLY=SHOULDER)
 *   SKIP_THUMBS  set to 1 to leave existing thumbnails alone (geometry-only re-bake)
 *
 * The converter used to be built from workers/cad-converter in the Cascadia-App
 * repo. That path does not exist here, so we pull the image the Cascadia-App CI
 * already publishes. Point CAD_IMAGE at a locally-built tag to test a converter
 * change before it lands.
 */

import { execFileSync, execSync, spawnSync } from 'node:child_process'
import { existsSync, mkdirSync, readFileSync, readdirSync } from 'node:fs'
import { dirname, join, resolve } from 'node:path'
import { fileURLToPath } from 'node:url'

const __dirname = dirname(fileURLToPath(import.meta.url))
const REPO_ROOT = resolve(__dirname, '..')
const DEMO_DIR = join(REPO_ROOT, 'robot-arm')
const STEP_DIR = join(DEMO_DIR, 'step')
const GLB_DIR = join(DEMO_DIR, 'glb')
const THUMB_DIR = join(DEMO_DIR, 'thumbnails')
const NODE_DIR = join(DEMO_DIR, 'nodes')
const SCRIPT_DIR = join(REPO_ROOT, 'scripts')
const PYTHON_SCRIPT = join(SCRIPT_DIR, '_prepare-demo-cad.py')

const IMAGE = process.env.CAD_IMAGE ?? 'ghcr.io/cascadia-plm/cascadia-cad-converter:latest'

function run(cmd: string, args: Array<string>, opts: { quiet?: boolean } = {}): void {
  if (!opts.quiet) console.log(`$ ${cmd} ${args.join(' ')}`)
  const result = spawnSync(cmd, args, { stdio: 'inherit' })
  if (result.status !== 0) {
    throw new Error(`Command failed: ${cmd} ${args.join(' ')}`)
  }
}

function imageExists(tag: string): boolean {
  try {
    const out = execFileSync('docker', ['image', 'inspect', tag], { stdio: 'pipe' })
    return out.length > 0
  } catch {
    return false
  }
}

// ----------------------------------------------------------------------------
// Pre-flight
// ----------------------------------------------------------------------------

if (!existsSync(STEP_DIR)) {
  console.error(`No step directory at ${STEP_DIR}. Run scripts/build-demo-manifest.ts first.`)
  process.exit(1)
}
if (!existsSync(PYTHON_SCRIPT)) {
  console.error(`Missing helper script: ${PYTHON_SCRIPT}`)
  process.exit(1)
}

const steps = readdirSync(STEP_DIR).filter((f) => f.toLowerCase().endsWith('.step')).sort()
if (steps.length === 0) {
  console.error(`No .step files in ${STEP_DIR}.`)
  process.exit(1)
}

console.log(`[prepare-demo-cad] ${steps.length} STEP files to process`)
console.log(`[prepare-demo-cad] image=${IMAGE}`)

mkdirSync(GLB_DIR, { recursive: true })
mkdirSync(THUMB_DIR, { recursive: true })
mkdirSync(NODE_DIR, { recursive: true })

// ----------------------------------------------------------------------------
// Pull the converter image if missing
// ----------------------------------------------------------------------------

if (!imageExists(IMAGE)) {
  console.log(`[prepare-demo-cad] pulling ${IMAGE} (this may take a few minutes)`)
  run('docker', ['pull', IMAGE])
} else {
  console.log(`[prepare-demo-cad] image already present, skipping pull`)
}

// ----------------------------------------------------------------------------
// Run the conversion in one container
// ----------------------------------------------------------------------------

// On Windows, Docker accepts both forms; using the absolute path that exists in WSL/Git-Bash.
const args: Array<string> = [
  'run', '--rm',
  '-v', `${DEMO_DIR}:/work`,
  '-v', `${SCRIPT_DIR}:/host-scripts:ro`,
  '-e', 'PYTHONUNBUFFERED=1',
  '-e', `FORCE=${process.env.FORCE ?? '0'}`,
  '-e', `ONLY=${process.env.ONLY ?? ''}`,
  '-e', `SKIP_THUMBS=${process.env.SKIP_THUMBS ?? '0'}`,
  '--entrypoint', 'sh',
  IMAGE,
  '-c',
  // Start Xvfb so render_thumbnail has a display, then run the Python helper.
  'Xvfb :99 -screen 0 512x512x24 -nolisten tcp >/dev/null 2>&1 & sleep 1; '
  + 'export DISPLAY=:99; '
  + 'exec python /host-scripts/_prepare-demo-cad.py /work',
]

console.log(`[prepare-demo-cad] starting container...`)
run('docker', args)

// ----------------------------------------------------------------------------
// Verify outputs
// ----------------------------------------------------------------------------

const glbs = readdirSync(GLB_DIR).filter((f) => f.endsWith('.glb'))
const thumbs = readdirSync(THUMB_DIR).filter((f) => f.endsWith('.png'))
const nodeFiles = readdirSync(NODE_DIR).filter((f) => f.endsWith('.json'))

// How many models the viewer will be able to take apart, and into how many
// parts. A model with no nodes is a single part, which is not a failure.
let structured = 0
let parts = 0
for (const f of nodeFiles) {
  const { nodes } = JSON.parse(readFileSync(join(NODE_DIR, f), 'utf-8')) as {
    nodes: Array<unknown>
  }
  if (nodes.length > 0) structured += 1
  parts += nodes.length
}

console.log()
console.log(`  STEP inputs:  ${steps.length}`)
console.log(`  GLB outputs:  ${glbs.length}`)
console.log(`  Thumbnails:   ${thumbs.length}`)
console.log(`  Selectable:   ${structured} assemblies, ${parts} parts`)

if (glbs.length === 0) {
  console.error('No GLBs produced. Check container logs above.')
  process.exit(1)
}
